### PR TITLE
fix: infer registry_name when MCP registry API returns empty values

### LINKED ISSUE #180: [BUG] All client adapters fail to resolve MCP registry packages when registry_name is empty
**Describe the bug**

All three client adapters (VS Code, Copilot CLI, Codex) fail to install stdio-based MCP servers from the registry when the API returns an empty `registry_name` field for packages. The adapters match package types using `package.get("registry_name")` directly, which always fails with empty strings — resulting in either a misleading error or a misconfigured server entry.

**To Reproduce**

1. Run `apm install` with a dependency on a registry MCP server that only provides stdio packages (e.g. `com.microsoft/azure`)
2. APM detects packages exist ("Supported package types: npm, pypi, docker") but fails to map any of them
3. Package identifiers appear empty in output; the server is not correctly configured

**Expected behavior**

APM should correctly identify package types (npm, pypi, docker, etc.) and generate valid runtime configurations even when the registry API returns empty `registry_name`.

**Environment**

- OS: macOS / Linux / Windows (all affected)
- Python Version: 3.13
- APM Version: latest main

**Observed registry API response**

Querying the MCP registry API for the Azure MCP server (`https://api.mcp.github.com/v0/servers/d3965c5a53be4f8bab7921b9d0511419`) returns packages with **empty `registry_name`** for all entries:

```json
{
  "packages": [
    {
      "name": "@azure/mcp",
      "version": "2.0.0-beta.24",
      "registry_name": "",
      "package_arguments": [
        {"type": "positional", "value": "server"},
        {"type": "positional", "value": "start"}
      ]
    },
    {
      "name": "msmcp-azure",
      "version": "2.0.0-beta.24",
      "registry_name": "",
      "package_arguments": [
        {"type": "positional", "value": "server"},
        {"type": "positional", "value": "start"}
      ]
    },
    {
      "name": "Azure.Mcp",
      "version": "2.0.0-beta.24",
      "registry_name": "",
      "package_arguments": []
    },
    {
      "name": "ghcr.io/azure/azure-mcp:latest",
      "version": "2.0.0-beta.24",
      "registry_name": "",
      "package_arguments": []
    }
  ]
}
```

This pattern is consistent across multiple servers in the registry — `registry_name` is empty for the majority of packages. The adapters' `_select_best_package()` and `_format_server_config()` methods rely on exact string matching against `registry_name`, causing all package type resolution to fail.

**Additional context**

- All three adapters are affected: `vscode.py`, `copilot.py`, `codex.py`
- The `package_arguments` field uses `{type, value}` format (e.g. `{"type": "positional", "value": "server"}`) which differs from the legacy `runtime_arguments` format (`{is_required, value_hint}`)
- Package types can be reliably inferred from: `runtime_hint` (e.g. `npx` → npm, `uvx` → pypi) and package name patterns (e.g. `@scope/name` → npm, `ghcr.io/…` → docker)

### PR DESCRIPTION
**Pull request by sergio-sisternes-epam**
## Description

The MCP registry API returns empty `registry_name` for the majority of packages, causing all three client adapters to fail when resolving stdio-based MCP servers. This PR adds registry type inference to each adapter so package selection and configuration work regardless of whether `registry_name` is populated.

Fixes #180

## Type of change

- [x] Bug fix
- [ ] New feature
- [x] Documentation
- [ ] Maintenance / refactor

### Root cause

The live MCP registry API (`api.mcp.github.com`) returns `"registry_name": ""` for most packages. All three adapters (`vscode.py`, `copilot.py`, `codex.py`) match on this field directly in `_select_best_package()` and `_format_server_config()`, so package type resolution always fails.

### Solution

Added `_infer_registry_name()` static method to each adapter that derives the package type through a fallback chain:

1. Explicit `registry_name` (when non-empty)
2. `runtime_hint` mapping (`npx`/`npm` → npm, `uvx`/`pip` → pypi, `docker` → docker, `dotnet` → nuget)
3. Package name patterns (`@scope/name` → npm, `ghcr.io/…` → docker, `PascalCase.Name` → nuget, `*.mcpb` → mcpb)

Additionally for the VS Code adapter:
- Added `_extract_package_args()` to normalize both `package_arguments` (current API: `{type, value}`) and `runtime_arguments` (legacy: `{is_required, value_hint}`)
- Added `-y` flag to npx commands for non-interactive installs

### Files changed

| File | Change |
|------|--------|
| `src/apm_cli/adapters/client/vscode.py` | Added `_infer_registry_name`, `_extract_package_args`; updated selection and formatting |
| `src/apm_cli/adapters/client/copilot.py` | Added `_infer_registry_name`; updated `_select_best_package` and registry_name usage |
| `src/apm_cli/adapters/client/codex.py` | Added `_infer_registry_name`; updated `_select_best_package` and registry_name usage |
| `tests/unit/test_vscode_adapter.py` | 15 new tests across 3 test classes (`TestVSCodeInferRegistryName`, `TestVSCodeExtractPackageArgs`, `TestVSCodeRealApiFormat`) |
| `tests/unit/test_runtime_args.py` | Updated assertions for `-y` flag in npx args |
| `tests/unit/test_env_variables.py` | Updated assertion for `-y` flag in npx args |
| `docs/integrations.md` | Added "Package Type Inference" section |

## Testing

- [x] Tested locally
- [x] All existing tests pass (1039 passed)
- [x] Added tests for new functionality (15 new tests including real API format coverage)

### CODE REVIEW

File: src/apm_cli/adapters/client/vscode.py [hunk 1]
```diff
@@ -189,40 +189,24 @@ def _format_server_config(self, server_info):
...
-                    args = [package.get("name")]
+            if runtime_hint == "npx" or registry_name == "npm":
+                args = pkg_args if pkg_args else [package.get("name")]
                 
                 server_config = {
                     "type": "stdio",
                     "command": "npx",
-                    "args": args
+                    "args": ["-y"] + args
                 }
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/adapters/client/vscode.py:208]: Fixed in 047426f. npx args now always include the package name: `npx -y @azure/mcp server start`. Legacy `runtime_arguments` that already contain the package name are deduplicated via filtering.

File: src/apm_cli/adapters/client/vscode.py [hunk 2]
```diff
@@ -231,42 +215,44 @@ def _format_server_config(self, server_info):
...
                 else:
-                    command = "python3"
-                
-                # Get args directly from runtime_arguments
-                args = []
-                if "runtime_arguments" in package and package["runtime_arguments"]:
-                    for arg in package["runtime_arguments"]:
-                        if arg.get("is_required", False) and arg.get("value_hint"):
-                            args.append(arg.get("value_hint"))
+                    command = "uvx"
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/adapters/client/vscode.py:229]: Won't fix — APM uses `uv` as its package manager, so `uvx` availability is a reasonable assumption. Defaulting to `python3` would break pypi packages designed to run via `uvx`. When `runtime_hint` explicitly says something else (e.g. `pip`, `python`), we already branch to `python3`.

File: tests/unit/test_vscode_adapter.py [hunk 3]
```diff
@@ -324,5 +324,405 @@ def test_configure_self_defined_http_via_cache(self, mock_get_path):
...
+
+        self.mock_integration_patcher = patch('apm_cli.adapters.client.vscode.RegistryIntegration')
+        self.mock_integration_patcher.start()
+
+    def tearDown(self):
+        import gc; gc.collect()
+        import time; time.sleep(0.1)
+        self.mock_registry_patcher.stop()
+        self.mock_integration_patcher.stop()
+        self.temp_dir.cleanup()
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply tests/unit/test_vscode_adapter.py:406]: Out of scope for this bug fix — the `gc.collect()` + `sleep` pattern is consistent with the existing test classes in this file. Will address in a separate cleanup PR.

File: tests/unit/test_vscode_adapter.py [hunk 4]
```diff
@@ -324,5 +324,405 @@ def test_configure_self_defined_http_via_cache(self, mock_get_path):
...
+
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply tests/unit/test_vscode_adapter.py:633]: Same as above — out of scope, will address in a separate cleanup PR.

File: src/apm_cli/adapters/client/vscode.py [hunk 5]
```diff
@@ -309,8 +295,133 @@ def _format_server_config(self, server_info):
...
+            return "npm"  # scoped npm package, e.g. @azure/mcp
+        if name.startswith(("ghcr.io/", "mcr.microsoft.com/", "docker.io/")):
+            return "docker"
+        if name.startswith("https://") and name.endswith(".mcpb"):
+            return "mcpb"
+        # PascalCase with dots usually means nuget (e.g. Azure.Mcp)
+        if "." in name and not name.startswith("http") and name[0].isupper():
+            return "nuget"
+        
+        return ""
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply src/apm_cli/adapters/client/vscode.py]: Fixed in 047426f — moved `_infer_registry_name()` to `MCPClientAdapter` base class in `base.py`. All three adapters now inherit it, eliminating the duplication.

File: tests/unit/test_vscode_adapter.py [hunk 6]
```diff
@@ -324,5 +324,405 @@ def test_configure_self_defined_http_via_cache(self, mock_get_path):
...
+        self.assertTrue(result)
+        with open(self.temp_path, "r") as f:
+            config = json.load(f)
+
+        server = config["servers"]["azure"]
+        self.assertEqual(server["type"], "stdio")
+        self.assertEqual(server["command"], "npx")
+        # npm inferred from @azure/mcp scoped name, -y flag added, pkg args appended
+        self.assertEqual(server["args"], ["-y", "server", "start"])
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply tests/unit/test_vscode_adapter.py]: Fixed in 047426f — tests now assert `["-y", "@azure/mcp", "server", "start"]`.

File: tests/unit/test_vscode_adapter.py [hunk 7]
```diff
@@ -324,5 +324,405 @@ def test_configure_self_defined_http_via_cache(self, mock_get_path):
...
+        )
+
+        self.assertTrue(result)
+        with open(self.temp_path, "r") as f:
+            config = json.load(f)
+
+        server = config["servers"]["azure"]
+        self.assertEqual(server["type"], "stdio")
+        self.assertEqual(server["command"], "npx")
+        self.assertEqual(server["args"], ["-y", "server", "start"])
```
[the remark this answers is not in the document]
Comment (sergio-sisternes-epam) [reply tests/unit/test_vscode_adapter.py]: Fixed in 047426f — tests now assert `["-y", "@azure/mcp", "server", "start"]`.

**Review by danielmeppiel — APPROVED** [review]
LGTM — most impactful of the three PRs. Fixes a real production bug where most MCP registry packages fail to install because `registry_name` is empty.

**What I verified:**
- `_infer_registry_name()` placed in `base.py` — correct, avoids triplication across adapters ✓
- Inference chain is solid: explicit → `runtime_hint` → name pattern heuristics ✓
- `-y` flag added to npx in VS Code + Codex adapters (Copilot adapter already had it) ✓
- `_extract_package_args()` normalizes both `package_arguments` (current API) and `runtime_arguments` (legacy) — future-proofs against API evolution ✓
- Generic `elif package and runtime_hint:` fallback in VS Code adapter catches nuget/dotnet and other runtimes ✓
- Error messages now list available registries instead of generic "no package information" — good UX ✓
- 15+ tests including real API format coverage with actual registry response shapes ✓
- `docs/integrations.md` updated ✓

**Non-blocking observations:**
1. **Pre-existing edge case**: If a package list exists but the selected package has no `runtime_hint` and doesn't match npm/docker/pypi, `server_config` would be undefined when the env var block tries to reference it. The new generic fallback reduces this risk significantly, but a `server_config = {}` initialization or an explicit `else` guard would be belt-and-suspenders. Pre-existing though, not introduced here.
2. VS Code `_select_best_package` priority is `["npm", "pypi", "docker"]` while Copilot/Codex use `["npm", "docker", "pypi", "homebrew"]`. This seems intentional but worth a comment explaining the rationale.

**Merge order:** PR 3 of 3. Rebase onto main after #178 merges.
